Fix default featured image size

derweili opened the pull request:
## Description
Fixed the default image sizes used for featured images displayed in the editor:
Fixes: #15842, #14084


## How has this been tested?
Verified with a linter on my local environment.


## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

derweili committed:
Change featured image default size to thumnail

jorgefilipecosta commented on packages/editor/src/components/post-featured-image/index.js:
```diff
@@ -31,7 +31,7 @@ function PostFeaturedImage( { currentPostId, featuredImageId, onUpdateImage, onR
 	let mediaWidth, mediaHeight, mediaSourceUrl;
 	if ( media ) {
-		const mediaSize = applyFilters( 'editor.PostFeaturedImage.imageSize', 'post-thumbnail', media.id, currentPostId );
+		const mediaSize = applyFilters( 'editor.PostFeaturedImage.imageSize', 'thumbnail', media.id, currentPostId );
```
Hi @derweili, thank you for your contribution 👍
I verified that the classic editor also loads post thumbnail when it is available.
E.g: this is the markup classic editor provided (the src is for the post-thumbnail size):
<img width="493" alt="Screenshot 2019-05-27 at 22 19 23" src="https://user-images.githubusercontent.com/11271197/58439798-09f23f80-80ce-11e9-9dad-5c0f43ae59ce.png">

I verified that if post_thumbnail is not available thumbnail size used in the classic editor while in Gutenberg full size is used. I think maybe we should follow the same logic verify if post_thumbnail is available and if yes we should use it, if not Gutenberg verifies if thumbnail is available.

swissspidy replied:
Makes sense to me. The `post-thumbnail` image size is available if the theme adds support for post thumbnails.

`thumbnail` on the other hand is always there, but is usually something like 150x150 big.

derweili replied:
@swissspidy `thumbnail` is not always there. 

If the uploaded image is smaller than `thumbnail` size (defaut 150 x 150) the thumbnail size is missing.
`full` is the only size that's always available.

I think that the image sizes should be chosen in the following order:
1. Load `post-thumbnail` (still use the `editor.PostFeaturedImage.imageSize` filter)
2. If `post-thumbnail` is not available, load `thumbnail`
3. If both `post-thumbnail` and `thumbnail` are not available, use the `full` size.

jorgefilipecosta replied:
Hi @derweili, that seems a good plan 👍

derweili replied:
@jorgefilipecosta I implemented the changes as described.
I reset the `mediaSize` to `post-thumbnail` and added a new `fallbackMediaSize` which has the value `thumbnail` by default.

I also added a new filter `editor.PostFeaturedImage.fallbackImageSize` to modify the new `fallbackMediaSize`, similar to the existing filter `editor.PostFeaturedImage.imageSize`

pablinos replied:
> `full` is the only size that's always available.

The frustrating thing here is that `media.media_details.sizes` is an empty object when the image is smaller than the `thumbnail` size. If it always contained a size then we could do something to pick the appropriate size in the order of preference. 

Given the current structure of the `media.media_details` object, this change looks good to me.

Alternatively, we could make another change so `sizes` always contains `full` at a minimum. 

Any thoughts?

jorgefilipecosta replied:
Thank you for the improvements @derweili this PR seems to be on a good path I left a suggestion.

> Alternatively, we could make another change so sizes always contains full at a minimum.

It would be good if originally that was the case but now it may be risky to change the structure as there may be external code/filters that is expecting and empty array.
I think the current direction addresses the problem well.

derweili committed:
fallbackMediaSize added to featured image

mediaSize reset to post-thumbnail
fallbackMediaSize added and set to thumbnail

`editor.PostFeaturedImage.fallbackImageSize` filter added to modify the new fallbackMediaSize

jorgefilipecosta commented on packages/editor/src/components/post-featured-image/index.js:
```diff
@@ -32,11 +32,20 @@ function PostFeaturedImage( { currentPostId, featuredImageId, onUpdateImage, onR
 	let mediaWidth, mediaHeight, mediaSourceUrl;
 	if ( media ) {
 		const mediaSize = applyFilters( 'editor.PostFeaturedImage.imageSize', 'post-thumbnail', media.id, currentPostId );
+		const fallbackMediaSize = applyFilters( 'editor.PostFeaturedImage.fallbackImageSize', 'thumbnail', media.id, currentPostId );
```
Do we really need a new filter for the fallback? Can't we use the existing filter e.g:
```
const fallbackMediaSize = applyFilters( 'editor.PostFeaturedImage. imageSize', 'thumbnail', media.id, currentPostId );
````
Another thing to consider is that the filters used here may have complex logic it would be nice to only compute the fallbackMediaSize if the mediaSize does not exist.

derweili replied:
@jorgefilipecosta 

I implemented you suggestions in the latest commit.

- The fallback size now uses the same filter as the default size.
- The fallback size now get computed only when default size is missing

jorgefilipecosta reviewed (commented):
Thank you for all the interactions @derweili I left a suggestion but other than that the PR seems to be very close to the finish line.

Commit:
PostFeaturedImage Performance optimizations

improved: the fallback size will now be calculated only if the default size is not available.
changed: the fallback size now uses the same filter as the default size

jorgefilipecosta reviewed (approved):
LGTM 👍
